### PR TITLE
Add Email template history DDL

### PR DESCRIPTION
**Pull request by tomrndom**
ref: https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskProject&view=vertical&task=3367287&stafilter=NotComplete&fileview=grid

Signed-off-by: Tomás Castillo <tcastilloboireau@gmail.com>

### CODE REVIEW

File: src/actions/email-actions.js [hunk 1]
```diff
@@ -95,7 +95,7 @@ export const getEmailTemplate = (templateId) => async (dispatch, getState) => {
 
     dispatch(startLoading());
 
-    const params = { access_token : accessToken, expand: 'parent' };
+    const params = { access_token : accessToken, expand: 'parent, versions' };
```
Comment (smarcet) [inline src/actions/email-actions.js:98]: remove space on expand, its not necessary

File: src/components/forms/email-template-form.js [hunk 2]
```diff
@@ -220,6 +221,20 @@ const EmailTemplateForm = ({ entity, match, errors, clients, preview, templateLo
...
 
+    const handleVersionChange = (ev) => {
+        const {value} = ev.target;
+        const selectedHistory = default_history.find(h => h.sha === value);
+        setHistoryVersion(selectedHistory.sha);
+        if(selectedHistory.type === 'html') {
+            setMjmlEditor(false);
+            setStateEntity({ ...stateEntity, html_content: selectedHistory.content });
+        }
+        if(selectedHistory.type === 'mjml') {
```
Comment (smarcet) [inline src/components/forms/email-template-form.js:232]: move this type to constants

File: src/components/forms/email-template-form.js [hunk 3]
```diff
@@ -338,6 +356,18 @@ const EmailTemplateForm = ({ entity, match, errors, clients, preview, templateLo
                                                     className={`btn btn-primary`} value={T.translate("emails.display_mjml")} />
                                             </>
                                         }
+                                        </div>
+                                        <div className='col-md-8'>
+                                            {entity.id > 0 && stateEntity.versions > 0 &&
```
Comment (smarcet) [inline src/components/forms/email-template-form.js:361]: its missing length here ?

File: src/components/forms/email-template-form.js [hunk 4]
```diff
@@ -228,8 +243,8 @@ const EmailTemplateForm = ({ entity, match, errors, clients, preview, templateLo
         };
     });
 
-
     const email_clients_ddl = clients ? clients.map(cli => ({ label: cli.name, value: cli.id })) : [];
+    const versions_ddl = stateEntity.versions ? stateEntity.versions.map(v => ({ label: v.last_modified, value: v.sha })) : [];
```
Comment (smarcet) [inline src/components/forms/email-template-form.js:247]: label per request should follow this format
last_modfied - sha1 - commit message
also it should have a link to open the html_url attribute to redirect to github ( opening on a new tab )

**Review by smarcet — CHANGES_REQUESTED** [review]
@tomrndom please review comments

**Comment by tomrndom** [comment]
Updated style and data for DDL

image

File: src/reducers/emails/email-template-list-reducer.js [hunk 5]
```diff
@@ -22,7 +22,7 @@ import { LOGOUT_USER } from 'openstack-uicore-foundation/lib/security/actions';
 
 const DEFAULT_STATE = {
     templates       : [],
-    term            : null,
+    term            : '',
```
Comment (smarcet) [inline src/reducers/emails/email-template-list-reducer.js:25]: why use empty ?
should be null

File: src/components/forms/email-template-form.js [hunk 6]
```diff
@@ -77,8 +77,18 @@ const EmailTemplateForm = ({ entity, match, errors, clients, preview, templateLo
...
         }
 
+        if (!shallowEqual(stateEntity, entity)) {
+            setStateEntity({ ...entity })
+        }
+
     }, [errors, entity]);
 
+    useEffect(() => {
+        const latestVersion = entity.versions[0];
```
Comment (smarcet) [inline src/components/forms/email-template-form.js:87]: this is not necessary why forcing this

File: src/actions/email-actions.js [hunk 7]
```diff
@@ -132,8 +132,10 @@ export const saveEmailTemplate = (entity, noAlert = false) => async (dispatch, g
             entity
         )(params)(dispatch)
             .then((payload) => {
-                if (!noAlert)
+                if (!noAlert){
                     dispatch(showSuccessMessage(T.translate("emails.template_saved")));
+                    dispatch(getEmailTemplate(entity.id));
```
Comment (smarcet) [inline src/actions/email-actions.js:137]: is not the payload already updated with the latest info ?
its really necessary to perform another request ?, i do think that the PUT returns the entity updated

File: src/reducers/emails/email-template-reducer.js [hunk 8]
```diff
@@ -91,7 +91,7 @@ const emailTemplateReducer = (state = DEFAULT_STATE, action) => {
 
             return {...state, entity: {...DEFAULT_ENTITY, ...entity,
                     original_mjml_content: entity.mjml_content,
-                    original_html_content:entity. html_content}, preview: null };
```
Comment (smarcet) [inline src/reducers/emails/email-template-reducer.js]: what was the purpose of the preview attribute ? why we need to remove it now ?

**Review by smarcet — CHANGES_REQUESTED** [review]
@tomrndom please review comments

File: src/actions/email-actions.js (hunk 7 above)
Comment (tomrndom) [reply src/actions/email-actions.js:137]: the PUT is missing the `versions` prop

Comment (smarcet) [reply src/actions/email-actions.js:137]: should be fixed
if u add this action 
https://github.com/fntechgit/summit-admin/blob/36064e9d7969ec5177d2bf690ff8521e1e6a884b/src/actions/email-actions.js#L40
on the reducer here[
](https://github.com/fntechgit/summit-admin/blob/36064e9d7969ec5177d2bf690ff8521e1e6a884b/src/reducers/emails/email-template-reducer.js#L82)
the the extra api call will not be necessary

Comment (smarcet) [reply src/actions/email-actions.js:137]: also recall to put the proper expand here
https://github.com/fntechgit/summit-admin/blob/36064e9d7969ec5177d2bf690ff8521e1e6a884b/src/actions/email-actions.js#L122
expand: 'parent,versions'

**Pull request merged**